optuna/terminator/improvement/emmr.py

```diff
@@ -0,0 +1,390 @@
+        min_n_trials:
+            A minimum number of complete trials to compute the criterion. Default to 2.
+        seed:
+            A random seed for EMMREvaluator.
+
+    Example:
+
+        .. testcode::
+
+            import logging
+
+            import optuna
+            from optuna.terminator import EMMREvaluator, MedianErrorEvaluator, Terminator
+
+            sampler = optuna.samplers.TPESampler(seed=0)
+            study = optuna.create_study(sampler=sampler, direction="minimize")
+            emmr_improvement_valuator = EMMREvaluator()
+            median_error_evaluator = MedianErrorEvaluator(emmr_improvement_valuator)
+            terminator = Terminator(
+                improvement_evaluator=emmr_improvement_valuator,
+                error_evaluator=median_error_evaluator,
+            )
+
+
+            for i in range(1000):
+                trial = study.ask()
+
+                ys = [trial.suggest_float(f"x{i}", -10.0, 10.0) for i in range(5)]
+                value = sum(ys[i] ** 2 for i in range(5))
+                logging.info(f"Trial #{trial.number} finished with value {value}.")
+
+                study.tell(trial, value)
+
+                if terminator.should_terminate(study):
+                    logging.info("Terminated by Optuna Terminator!")
+                    break
+            else:
+                logging.info("Not terminated.")
+
+    """
+
+    def __init__(
+        self,
+        deterministic_objective: bool = False,
+        delta: float = 0.1,
+        min_n_trials: int = 2,
+        seed: int | None = None,
+    ) -> None:
+        if min_n_trials <= 1 or not np.isfinite(min_n_trials):
+            raise ValueError("`min_n_trials` is expected to be a finite integer more than one.")
+
+        self._deterministic = deterministic_objective
+        self._delta = delta
+        self.min_n_trials = min_n_trials
+        self._rng = LazyRandomState(seed)
+
+    def evaluate(self, trials: list[FrozenTrial], study_direction: StudyDirection) -> float:
+
+        optuna_search_space = intersection_search_space(trials)
+        complete_trials = [t for t in trials if t.state == TrialState.COMPLETE]
+
+        if len(complete_trials) < self.min_n_trials:
+            return sys.float_info.max  # Do not terminate.
+
+        search_space, normalized_params = gp_search_space.get_search_space_and_normalized_params(
+            complete_trials, optuna_search_space
+        )
+        if len(search_space.scale_types) == 0:
+            warnings.warn(
+                f"{self.__class__.__name__} cannot consider any search space."
+                "Termination will never occur in this study."
+            )
+            return sys.float_info.max  # Do not terminate.
+
+        len_trials = len(complete_trials)
+        len_params = len(search_space.scale_types)
+        assert normalized_params.shape == (len_trials, len_params)
+
+        # _gp module assumes that optimization direction is maximization
```

Comment:
A nit.
```suggestion
        # _gp module assumes that optimization direction is maximization.
```

Reply:
Please refer to the [code convention](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences) of this repository.